value/structreflect.go

```diff
@@ -0,0 +1,238 @@
+func (c *structCache) Get(t reflect.Type) (map[string]*fieldCacheEntry, bool) {
+	entry, ok := c.value.Load().(structCacheMap)[t]
+	return entry, ok
+}
+
+// Update sets the fieldCacheEntry for the given type via a copy-on-write update to the struct cache.
+func (c *structCache) Update(t reflect.Type, m map[string]*fieldCacheEntry) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	oldCacheMap := c.value.Load().(structCacheMap)
+	newCacheMap := make(structCacheMap, len(oldCacheMap)+1)
+	for k, v := range oldCacheMap {
+		newCacheMap[k] = v
+	}
+	newCacheMap[t] = m
+	c.value.Store(newCacheMap)
+}
+
+func newStructCache() *structCache {
+	cache := &structCache{}
+	cache.value.Store(make(structCacheMap))
+	return cache
+}
+
+type fieldCacheEntry struct {
+	// isOmitEmpty is true if the field has the json 'omitempty' tag.
+	isOmitEmpty bool
+	// fieldPath is the field indices (see FieldByIndex) to lookup the value of
+	// a field in a reflect.Value struct. A path of field indices is used
+	// to support traversing to a field nested in struct fields that have the 'inline'
+	// json tag.
+	fieldPath [][]int
+}
+
+func (f *fieldCacheEntry) getFieldFromStruct(structVal reflect.Value) reflect.Value {
+	// field might be nested within 'inline' structs
+	for _, elem := range f.fieldPath {
+		structVal = structVal.FieldByIndex(elem)
+	}
+	return structVal
+}
+
+func getStructCacheEntry(t reflect.Type) structCacheEntry {
+	if hints, ok := reflectStructCache.Get(t); ok {
+		return hints
+	}
+
+	hints := map[string]*fieldCacheEntry{}
+	buildStructCacheEntry(t, hints, nil)
+
+	reflectStructCache.Update(t, hints)
+	return hints
+}
+
+func buildStructCacheEntry(t reflect.Type, infos map[string]*fieldCacheEntry, fieldPath [][]int) {
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		jsonName, isInline, isOmitempty := lookupJsonTags(field)
+		if isInline {
+			buildStructCacheEntry(field.Type, infos, append(fieldPath, field.Index))
+			continue
+		}
+		info := &fieldCacheEntry{isOmitEmpty: isOmitempty, fieldPath: append(fieldPath, field.Index)}
+		infos[jsonName] = info
+
+	}
+}
+
+type structReflect struct {
+	Value reflect.Value
+}
+
+func (r structReflect) Length() int {
+	i := 0
+	eachStructField(r.Value, func(s string, value reflect.Value) bool {
+		i++
+		return true
+	})
+	return i
+}
+
+func (r structReflect) Get(key string) (Value, bool) {
+	if val, ok := r.findJsonNameField(key); ok {
+		return mustWrapValueReflect(val), true
+	}
+	return nil, false
+}
+
+func (r structReflect) Has(key string) bool {
+	_, ok := r.findJsonNameField(key)
+	return ok
+}
+
+func (r structReflect) Set(key string, val Value) {
+	fieldVal, ok := r.findJsonNameField(key)
+	if !ok {
+		panic(fmt.Sprintf("key %s may not be set on struct %T: field does not exist", key, r.Value.Interface()))
+	}
+	if !fieldVal.CanSet() {
+		// See https://blog.golang.org/laws-of-reflection for details on why a struct may not be settable
```

Comment:
In other reflection stuff, we require that a user pass a pointer for the top level object.

The other case, a struct as a map item, we can actually handle (with a lot of complexity) by storing the map and the key and copying and re-setting the whole item. Hopefully we don't need to do that.

I'm not sure if panic is going to be OK here unless we're sure it's not going to happen at run time?

Reply:
The value interfaces are read-only with the exception of Map,  which is currently call exclusively by `discriminator.Set` for noramlization. This set function feels out-of-place to me for this interface. I'll sort out what we need to do about this case tomorrow.
^ @apelisse

Reply:
Looks like we do always expect a pointer for the top level object. I'm going add some checks to enforce that, but looks like we're okay.

Reply:
Added a `TestReflectMutateNestedStruct` test to cover these cases. The tests failed when a struct being changed is contained in a map, so I've created a workaround that puts a replacement into the map.